system/jlib/jiface.hpp

```diff
@@ -119,31 +119,46 @@ class CInterfaceOf : public CSimpleInterfaceOf<INTERFACE>
 public:
     virtual void beforeDispose() {}
 
-    inline bool isAlive() const         { return this->xxcount.load(std::memory_order_relaxed) < DEAD_PSEUDO_COUNT; }       //only safe if Link() is called first
+    //Function to be called when checking if an entry in a non-linking cache is valid
+    //must be called inside a critical section that protects access to the cache.
+    //This function must only increment the link count if the object is valid - otherwise it
+    //end up being freed more than once.
+    inline bool isAliveAndLink() const
+    {
+        unsigned expected = this->xxcount.load(std::memory_order_acquire);
+        for (;;)
+        {
+            //If count is 0, or count >= DEAD_PSEUDO_COUNT then return false - combine both into a single test
+            if ((expected-1) >= (DEAD_PSEUDO_COUNT-1))
+                return false;
+
+            //Because beforeDispose could cause this object to be linked/released or call isAlive(), this->xxcount is set
+            //to a a high mid-point positive number to avoid poss. of releasing again.
+            if (this->xxcount.compare_exchange_weak(expected, expected+1, std::memory_order_acq_rel))
+                return true;
+        }
+    }
 
     inline bool Release(void) const
     {
         if (this->xxcount.fetch_sub(1,std::memory_order_release) == 1)
         {
-            unsigned zero = 0;
-            //Because beforeDispose could cause this object to be linked/released or call isAlive(), this->xxcount is set
-            //to a a high mid-point positive number to avoid poss. of releasing again.
-            if (this->xxcount.compare_exchange_strong(zero, DEAD_PSEUDO_COUNT, std::memory_order_acq_rel))
+            //Overwrite with a special value so that any calls to Link()/Release within beforeDispose()
+            //do not cause the object to be released again.
+            this->xxcount.store(DEAD_PSEUDO_COUNT, std::memory_order_release);
```

Comment:
I think I like the old method here - if count is zero then atomically exchange it with DEAD_PSEUDO_COUNT and call beforeDispose().
What prevents someone else from linking between the fetch_sub() and store() ?

Reply:
I think if LINK() is never called on it, but only isAliveAndLink() then its ok.

Reply:
If link is called, then you are back to the problem of having a potential race condition.

Reply:
and cas is much more expensive than a store.